wgsl/scanner.cc

```diff
@@ -0,0 +1,992 @@
+      // TODO(dneto): Same here, need to be able to un-advance to before '*'
+      } else if (match('*') && match('/')) {
+        nesting--;
+      } else {
+        next();
+      }
+    }
+    return true;
+  }
+
+  /// Advances the lexer while the next code point is considered whitespace
+  void skip_whitespace() {
+    while (is_space(peek())) {
+      lexer_->advance(lexer_, /* whitespace */ true);
+    }
+  }
+
+ private:
+  TSLexer* lexer_;
+};
+
+struct Scanner {
+  struct State {
+    BitQueue<1024> lt_is_tmpl;  // Queue of disambiguated '<'
+    BitQueue<1024> gt_is_tmpl;  // Queue of disambiguated '>'
+    bool empty() const { return lt_is_tmpl.empty() && gt_is_tmpl.empty(); }
+  };
+  State state;
+  static_assert(sizeof(State) < TREE_SITTER_SERIALIZATION_BUFFER_SIZE);
+  // State is trivially copyable, so it can be serialized and deserialized
+  // with memcpy.
+  static_assert(std::is_trivially_copyable<State>::value);
+
+  /// Updates #state with the disambiguated '<' and '>' tokens.
+  /// The following assumptions are made on entry:
+  /// * lexer has just advanced to the end of an identifier
+  /// On exit, all '<' and '>' template tokens will be paired up to the closing
+  /// '>' for the first '<'.
+  void classify_template_args(Lexer& lexer) {
+    LOG("classify_template_args()");
+
+    if (!lexer.match('<')) {
+      LOG("  missing '<'");
+      return;
+    }
+
+    // The current expression nesting depth.
+    size_t expr_depth = 0;
+
+    // A stack of '<' tokens.
+    // Used to pair '<' and '>' tokens at the same expression depth.
+    struct StackEntry {
+      size_t index;       // Index of the opening '>' in lt_is_tmpl
+      size_t expr_depth;  // The value of 'expr_depth' for the opening '<'
+    };
+    std::vector<StackEntry> lt_stack;
+
+    LOG("classify_template_args() '<'");
+    lt_stack.push_back(StackEntry{state.lt_is_tmpl.count(), expr_depth});
+    state.lt_is_tmpl.push_back(false);  // Default to less-than
+
+    while (!lt_stack.empty() && !lexer.match(kEOF)) {
+      lexer.skip_whitespace();
+
+      // TODO: skip line-ending comments.
+      if (lexer.match_block_comment()) {
+        continue;
+      }
+
+      if (lexer.match_identifier()) {
+        lexer.skip_whitespace(); // TODO: Skip comments
+        if (lexer.match('<')) {
+          LOG("classify_template_args() '<'");
+          lt_stack.push_back(StackEntry{state.lt_is_tmpl.count(), expr_depth});
+          state.lt_is_tmpl.push_back(false);  // Default to less-than
+        }
+        continue;
+      }
+
+      if (lexer.match('>')) {
+        LOG("classify_template_args() '>'");
+        if (!lt_stack.empty() && lt_stack.back().expr_depth == expr_depth) {
+          LOG("   TEMPLATE MATCH");
+          state.gt_is_tmpl.push_back(true);
+          state.lt_is_tmpl[lt_stack.back().index] = true;
+          lt_stack.pop_back();
+        } else {
+          LOG("   non-template '>'");
+          state.gt_is_tmpl.push_back(false);
```

Comment:
This is necessary to be able to parse a '>' as greater than, for example:

       var v: array<f32,select(1,2,(a>b))>;  // that a>b must a GREATER_THAN B